SimPEG/potential_fields/magnetics/simulation.py

```diff
@@ -517,306 +570,235 @@ def getB0(self):
-                \left[
-                    \Div \diag(\M^f_{\mu_{0}^{-1}}\mathbf{B}_0) \dMfMuI
-                \right]
-                - \diag(\mathbf{v}) \mathbf{D} \mathbf{P}_{out}^T
-                    \frac{\partial B_{sBC}}{\partial \mathbf{m}}
+        B, u = f["B"], f["u"]
 
-        In the end,
+        Q = self.projectFieldsDeriv(B)
+        B0 = self.getB0()
+        C = -self.MfMuiI * self._Div.T
 
-        .. math ::
+        db_dm = 0
+        dCmu_dm = 0
 
-            \frac{\delta \mathbf{u}}{\delta \mathbf{m}} =
-            - [ \mathbf{A} ]^{-1}
-            \left[
-                \frac{\partial \mathbf{A}}{\partial \mathbf{m}}(\mathbf{m})\mathbf{u}
-                - \frac{\partial \mathbf{rhs}(\mathbf{m})}{\partial \mathbf{m}}
-            \right]
+        mu = np.ones(self.mesh.n_cells) * self.mu
+        mu_vec = np.hstack((mu, mu, mu))
 
-        A little tricky point here is we are not interested in potential (u), but interested in magnetic flux (B).
-        Thus, we need sensitivity for B. Now we take derivative of B w.r.t m and have
+        if self.remMap is not None:
+            Mf_rem_deriv = self.mesh.get_face_inner_product_deriv(
+                np.ones(self.mesh.n_cells * 3)
+            )(np.ones(self.mesh.n_faces)) * sp.diags(1 / mu_vec) * self.remDeriv
+            db_dm += (self.MfMuiI * Mf_rem_deriv * v)
 
-        .. math ::
+        if self.muMap is not None:
+            MfMuiIderiv = self.MfMuiIDeriv
+            dCmu_dm += MfMuiIderiv(self._Div.T @ u, v, adjoint=False)
+            db_dm += self._MfMu0i * MfMuiIderiv(B0, v, adjoint=False)
 
-            \frac{\delta \mathbf{B}} {\delta \mathbf{m}} =
-            \frac{\partial \mathbf{\mu} } {\partial \mathbf{m} }
-            \left[
-                \diag(\M^f_{\mu_{0}^{-1} } \mathbf{B}_0) \dMfMuI  \
-                 - \diag (\Div^T\mathbf{u})\dMfMuI
-            \right ]
+            if getattr(self, "rem", None) is not None:
+                Mf_r_over_uvec = self.mesh.get_face_inner_product(self.rem / mu_vec).diagonal()
+                mu_vec_i_deriv = sp.vstack((self.muiDeriv, self.muiDeriv, self.muiDeriv))
+                Mf_mu_vec_i_deriv = self.mesh.get_face_inner_product_deriv(
+                    np.ones(self.mesh.n_cells * 3)
+                )(np.ones(self.mesh.n_faces)) * sp.diags(self.rem) * mu_vec_i_deriv
+                db_dm += MfMuiIderiv(Mf_r_over_uvec, v, adjoint=False) + (self.MfMuiI * Mf_mu_vec_i_deriv * v)
 
-             -  (\MfMui)^{-1}\Div^T\frac{\delta\mathbf{u}}{\delta \mathbf{m}}
+        dq_dm_min_dAmu_dm = self._Div * (-dCmu_dm + db_dm)
 
-        Finally we evaluate the above, but we should remember that
+        Ainv_dqmindAmu = self._Ainv * dq_dm_min_dAmu_dm
 
-        .. note ::
+        Jv = Q * (C * Ainv_dqmindAmu + (-dCmu_dm + db_dm))
 
-            We only want to evaluate
+        return Jv
 
-            .. math ::
+    def Jtvec(self, m, v, f=None):
```

Comment:
There's a lot of repeated code between the `Jtvec` function and the `getJ` function. You can likely combine them by making `v` a keyword argument to `Jtvec` with `v=None` and doing near the beginning:
```
if v is not None:
    QTv = Q.T @ v
else:
    QTv = Q.T.toarray()
```

Then `getJ` would just call `Jtvec(m, v=None, f=f)` to create the matrix.